### DISCUSSION TITLE
关于ClusterBasic.ResFlushInterval参数影响content-type类型的问题

### DISCUSSION THREAD

**Discussion by godcavalry** [bfenetworks/bfe#1030] [Q&A (问题和回答)] [answered]
问题如下：
- 当ResFlushInterval保持默认值-1不变时
    backend的response header中不添加content-type时，经过BFE调度后BFE会自动将Content-Type设置为 “text/plain; charset=utf-8”
- 当ResFlushInterval设置为非-1时
   backend的response header中不添加content-type时，经过BFE调度后BFE会自动根据数据类型添加对应的Content-Type

而我们期望的是经过BFE调度后BFE会自动根据数据类型添加对应的Content-Type,而不是统一全部设置为“text/plain; charset=utf-8”

**Comment by iyangsj** [comment, marked as answer]
**建议：**
- 建议在生成响应时，业务应正确设置Content-Type头部（例如在Web开发框架中统一支持）
- 建议保持使用ResFlushInterval默认值，后续版本该参数可能被废除

**原因：**
- 在接入层根据响应内容识别Content-Type，并添加缺失的相应头部存在一些副作用
  * 例如：使用HTTP长轮训模式推送消息，可能导致交互出现异常
- ResFlushInterval在配置非-1时，对性能的影响也比较明显
  * 例如：配置的间隔较小且并发刷新非常多时

**Reply by godcavalry** [reply]
我们肯定都希望所有的程序都设置正确的Content-Type, 但是事与愿违，有些业务确实是没有设置Content-Type，而nginx又能返回正确的Content-Type.这种情况就会阻碍我们不能将业务从Nginx迁移到bfe。

根据分析bfe源码，如果不启用缓存的情况下，“backend的response header中不添加content-type时，BFE调度后BFE会自动将Content-Type设置为 “text/plain; charset=utf-8”” 这个问题点我们定位到是 copyResponse方法中由于复制数据前先调用了FLush方法，而Flush方法会给根据nil数据自动设置content-type为text/plain; charset=utf-8。

经过测试，如果flush方法不设置content-type,则能返回正确的content-type。

**Reply by iyangsj** [reply]
@godcavalry 感谢你反馈的问题场景：
- 目前的bfe版本，可以设置例如 ResFlushInterval=100ms 来处理响应缺失Content-Type问题
- 后续的bfe版本考虑增加扩展模块，来单独处理Content-Type识别及添加问题，使用方可按需启用
（是否启用添加Content-Type，与是否Flush与将会解耦合）

**Comment by daimg** [comment]
@godcavalry  请关注#1032，它将能解决ResFlushInterval为默认-1时，设置Content-Type问题。
